### PR TITLE
[SPARK-17793] [Web UI] Sorting on the description on the Job or Stage page doesn’t always work

### PR DESCRIPTION
**Pull request by ajbozarth**
## What changes were proposed in this pull request?

Added secondary sorting on stage name for the description column. This provide a clearer behavior in the common case where the Description column only comprises of Stage names instead of the option description value.
## How was this patch tested?

manual testing and dev/run-tests

Screenshots of sorting on both description and stage name as well as an example of both:
screen shot 2016-10-04 at 1 09 39 pm
screen shot 2016-10-04 at 1 09 51 pm
screen shot 2016-10-05 at 1 14 45 pm
screen shot 2016-10-05 at 1 14 51 pm
screen shot 2016-10-05 at 4 42 52 pm

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 1]
```diff
@@ -462,7 +462,10 @@ private[ui] class JobDataSource(
       }
       case "Description" => new Ordering[JobTableRowData] {
         override def compare(x: JobTableRowData, y: JobTableRowData): Int =
-          Ordering.String.compare(x.lastStageDescription, y.lastStageDescription)
+          (x.lastStageDescription, y.lastStageDescription) match {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: You know, this and all the other Orderings here could be simpler. This one can be

`case "Description" => Ordering.by(x => (x.lastStageName, x.lastStageDescription))`

Others like the one above should be `Ordering.by(_.jobData.jobId)`. If that's right I think it may be useful to simplify this pattern everywhere while here.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala [hunk 2]
```diff
@@ -498,8 +501,7 @@ private[ui] class JobPagedTable(
     sortColumn: String,
     desc: Boolean
   ) extends PagedTable[JobTableRowData] {
-  val parameterPath = UIUtils.prependBaseUri(basePath) + s"/$subPath/?" +
-    parameterOtherTable.mkString("&")
+  val parameterPath = basePath + s"/$subPath/?" + parameterOtherTable.mkString("&")
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Is this related?

**Review by srowen — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 2 above)
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: Peripherally, `UIUtils.prependBaseUri()` is called on `basePath` before it's passed here, so this is redundant and breaks all the sorting links that rely on `parameterPath` when using yarn. I fixed it here since it was breaking my fix on yarn.

File: core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala (hunk 1 above)
Comment (ajbozarth) [inline core/src/main/scala/org/apache/spark/ui/jobs/AllJobsPage.scala]: I knew this function seemed bulky, I'll go through and update it as you suggested.

**Comment by ajbozarth** [comment]
Jenkins, retest this please

**Review by srowen — APPROVED** [review]

**Comment by ajbozarth** [comment]
Jenkins, retest this please

**Review by srowen — APPROVED** [review]
Looks good. That is so much simpler.

**Comment by ajbozarth** [comment]
Jenkin, retest this please

**Comment by ajbozarth** [comment]
It's failed twice now on the same issue but it's completely unrelated to my change. Trying one more time, but is this failure a known issue?

**Comment by ajbozarth** [comment]
Jenkins, retest this please

**Comment by srowen** [comment]
Merged to master
